File: ios/MindboxJsDelivery.m
//
//  MindboxJsDelivery.m
//  MindboxSdk
//
//  Created by Nikolay Seleznev on 15.10.2021.
//  Copyright © 2021 Facebook. All rights reserved.
//

#import "MindboxJsDelivery.h"
#import <UserNotifications/UserNotifications.h>

@implementation MindboxJsDelivery

RCT_EXPORT_MODULE();

static bool hasListeners = NO;
static NSDictionary *storedEventDetails;

- (NSArray<NSString *> *)supportedEvents {
  return @[@"pushNotificationClicked", @"Click", @"Dismiss"];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)startObserving {
  hasListeners = YES;

  [[NSNotificationCenter defaultCenter] addObserver:self
                                           selector:@selector(emitEventInternal:)
                                               name:@"event-emitted"
                                             object:nil];

  [[NSNotificationCenter defaultCenter] addObserver:self
                                           selector:@selector(inappActionReceived:)
                                               name:@"MindboxInappAction"
                                             object:nil];

  if (storedEventDetails != NULL) {
    [[NSNotificationCenter defaultCenter] postNotificationName:@"event-emitted"
                                                        object:self
                                                      userInfo:storedEventDetails];
  }
}

- (void)stopObserving {
  hasListeners = NO;

  [[NSNotificationCenter defaultCenter] removeObserver:self];

  if (storedEventDetails != NULL) {
    storedEventDetails = NULL;
  }
}

- (void)emitEventInternal:(NSNotification *)notification {
  NSArray *eventDetails = [notification.userInfo valueForKey:@"detail"];
  NSString *eventName = [eventDetails objectAtIndex:0];
  NSString *actionIdentifier = [eventDetails objectAtIndex:1];
  NSDictionary *userInfo = [eventDetails objectAtIndex:2];
  NSString *clickUrl = @"";
  NSString *pushPayload = @"";

  if ([actionIdentifier isEqual:UNNotificationDefaultActionIdentifier]) {
    clickUrl = [userInfo objectForKey:@"clickUrl"];
    pushPayload = [userInfo objectForKey:@"payload"];

    if ([clickUrl length] == 0) {
      NSDictionary *aps = [userInfo objectForKey:@"aps"];
      clickUrl = [aps objectForKey:@"clickUrl"];
      pushPayload = [aps objectForKey:@"payload"];
    }
  } else {
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"uniqueKey == %@", actionIdentifier];
    NSArray *filteredArray = [[userInfo objectForKey:@"buttons"] filteredArrayUsingPredicate:predicate];

    if (filteredArray.count > 0) {
      clickUrl = [filteredArray.firstObject objectForKey:@"url"];
    }

    pushPayload = [userInfo objectForKey:@"payload"];
    if ([pushPayload length] == 0) {
      NSDictionary *aps = [userInfo objectForKey:@"aps"];
      pushPayload = [aps objectForKey:@"payload"];
    }
  }

  NSDictionary *dict = @{
    @"pushUrl": clickUrl ? clickUrl : [NSNull null],
    @"pushPayload": pushPayload ? pushPayload : [NSNull null]
  };
  NSError *error;
  NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dict options:NSJSONWritingPrettyPrinted error:&error];
  NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];

  [self sendEventWithName:eventName body:jsonString];

  if (storedEventDetails != NULL) {
    storedEventDetails = NULL;
  }
}

+ (void)emitEvent:(UNNotificationResponse *)response {
  NSDictionary *userInfo = response.notification.request.content.userInfo;
  NSString *name = @"pushNotificationClicked";
  NSString *actionIdentifier = response.actionIdentifier;
  NSDictionary *eventDetails = @{@"detail":@[name,actionIdentifier,userInfo]};
  if (hasListeners) {
    [[NSNotificationCenter defaultCenter] postNotificationName:@"event-emitted" object:self userInfo:eventDetails];
  } else {
    storedEventDetails = eventDetails;
  }
}

+ (void)sendInappEvent:(NSString *)eventName eventId:(NSString *)eventId url:(NSString *)clickUrl payload:(NSString *)payload {
    if (hasListeners) {
        NSMutableDictionary *bodyDict = [[NSMutableDictionary alloc] init];
        bodyDict[@"id"] = eventId;

        if (clickUrl != nil) {
            bodyDict[@"redirectUrl"] = clickUrl;
        }

        if (payload != nil) {
            bodyDict[@"payload"] = payload;
        }

        NSDictionary *userInfo = @{
            @"eventName": eventName,
            @"body": bodyDict
        };

        [[NSNotificationCenter defaultCenter] postNotificationName:@"MindboxInappAction"
                                                            object:self
                                                          userInfo:userInfo];
    }
}

- (void)inappActionReceived:(NSNotification *)notification {
    NSString *eventName = notification.userInfo[@"eventName"];
    NSDictionary *body = notification.userInfo[@"body"];
    [self sendEventWithName:eventName body:body];
}

@end

File: ios/MindboxJsDelivery.swift
import Foundation
import UserNotifications

@objc(MindboxJsDelivery)
public final class MindboxJsDelivery: NSObject {

    @objc public static func emitEvent(_ response: UNNotificationResponse) {
        let userInfo = response.notification.request.content.userInfo
        let actionIdentifier = response.actionIdentifier
        let pushUrl = resolvePushUrl(userInfo: userInfo, actionIdentifier: actionIdentifier)
        let pushPayload = resolvePushPayload(userInfo: userInfo)
        MindboxSdkImpl.emitPushClick(pushUrl: pushUrl, pushPayload: pushPayload)
    }

    @objc public static func sendInappEvent(_ eventName: String, eventId: String, url: String?, payload: String?) {
        switch eventName {
        case "Click":
            MindboxSdkImpl.emitInAppClick(id: eventId, redirectUrl: url, payload: payload)
        case "Dismiss":
            MindboxSdkImpl.emitInAppDismiss(id: eventId)
        default:
            break
        }
    }

    private static func resolvePushUrl(userInfo: [AnyHashable: Any], actionIdentifier: String) -> String {
        if actionIdentifier == UNNotificationDefaultActionIdentifier {
            return readString(userInfo: userInfo, key: "clickUrl")
                ?? readString(userInfo: readDictionary(userInfo: userInfo, key: "aps"), key: "clickUrl")
                ?? ""
        }
        let buttonUrl = readButtonUrl(userInfo: userInfo, uniqueKey: actionIdentifier)
        if let buttonUrl = buttonUrl, !buttonUrl.isEmpty {
            return buttonUrl
        }
        let aps = readDictionary(userInfo: userInfo, key: "aps")
        return readButtonUrl(userInfo: aps, uniqueKey: actionIdentifier) ?? ""
    }

    private static func resolvePushPayload(userInfo: [AnyHashable: Any]) -> String {
        return readString(userInfo: userInfo, key: "payload")
            ?? readString(userInfo: readDictionary(userInfo: userInfo, key: "aps"), key: "payload")
            ?? ""
    }

    private static func readButtonUrl(userInfo: [AnyHashable: Any], uniqueKey: String) -> String? {
        guard let buttons = userInfo["buttons"] as? [[String: Any]] else {
            return nil
        }
        return buttons.first(where: { ($0["uniqueKey"] as? String) == uniqueKey })?["url"] as? String
    }

    private static func readDictionary(userInfo: [AnyHashable: Any], key: String) -> [AnyHashable: Any] {
        guard let dictionary = userInfo[key] as? [AnyHashable: Any] else {
            return [:]
        }
        return dictionary
    }

    private static func readString(userInfo: [AnyHashable: Any], key: String) -> String? {
        return userInfo[key] as? String
    }
}
